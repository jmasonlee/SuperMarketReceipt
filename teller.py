from model_objects import Offer
from receipt import Receipt


class Teller:

    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    def add_special_offer(self, offer_type, product, argument):
        self.offers[product] = Offer(offer_type, product, argument)

    def checks_out_articles_from(self, the_cart, receipt=None):
        #Begin Common Code
        # receipt = self.calculate_total_charges(the_cart)
        assert receipt is not None

        expected_receipt = Receipt()
        for item in the_cart.items:
            expected_receipt.add_product(item.product, item.quantity, self.catalog.unit_price(item.product), item.quantity * self.catalog.unit_price(item.product))
        assert str(receipt.items) == str(expected_receipt.items)
        ### EXIT CRITERION: receipt is not present
        ### EXIT CRITERION: receipt needs to have values for the items being purchased
        #End Common Code Note: handle_offers could also contain common code

        the_cart.handle_offers(receipt, self.offers, self.catalog)

        return receipt

    def calculate_total_charges(self, the_cart):
        receipt = Receipt()
        product_quantities = the_cart.items
        for pq in product_quantities:
            p = pq.product
            quantity = pq.quantity
            unit_price = self.catalog.unit_price(p)
            price = quantity * unit_price
            # price = 0.00 #quantity * unit_price
            receipt.add_product(p, quantity, unit_price, price)
        return receipt
