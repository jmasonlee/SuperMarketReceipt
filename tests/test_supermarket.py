import pytest

from model_objects import Product, SpecialOfferType, ProductUnit, Discount
from receipt import Receipt
from shopping_cart import ShoppingCart
from teller import Teller
from tests.fake_catalog import FakeCatalog
from approvaltests import verify


def test_ten_percent_discount():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush, 10.0)

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 2.5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = Receipt()
    receipt.add_product(apples, 2.5, 1.99, 4.975)
    receipt.add_product(toothbrush, 1, 0.99, 0.99)

    receipt = teller.checks_out_articles_from(cart, receipt)

    verify(str(receipt))


def test_full_receipt_no_discount():
    apples, teller, toothbrush = create_teller_and_items()

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 2.5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = Receipt()
    receipt.add_product(apples, 2.5, 1.99, 4.975)
    receipt.add_product(toothbrush, 1, 0.99, 0.99)

    receipt = teller.checks_out_articles_from(cart, receipt)

    verify(str(receipt))


def test_full_receipt_add_item_adds_one_item():
    apples, teller, toothbrush = create_teller_and_items()

    cart = ShoppingCart()
    cart.add_item(apples)
    cart.add_item(toothbrush)

    receipt = Receipt()
    receipt.add_product(apples, 1.0, 1.99, 1.99)
    receipt.add_product(toothbrush, 1.0, 0.99, 0.99)

    receipt = teller.checks_out_articles_from(cart, receipt)

    verify(str(receipt))


def test_product_quantities_updates():
    apples, teller, toothbrush = create_teller_and_items()

    cart = ShoppingCart()
    cart.add_item(apples)
    cart.add_item_quantity(toothbrush, 2.0)
    cart.add_item(toothbrush)

    receipt = Receipt()
    receipt.add_product(apples, 1.0, 1.99, 1.99)
    receipt.add_product(toothbrush, 2.0, 0.99, 1.98)
    receipt.add_product(toothbrush, 1.0, 0.99, 0.99)

    receipt = teller.checks_out_articles_from(cart, receipt)
    verify(str(cart.product_quantities) + str(receipt))


def test_full_receipt_three_for_two():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.THREE_FOR_TWO, toothbrush, "foor")

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 2.5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def test_full_receipt_two_for_one():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.TWO_FOR_AMOUNT, apples, 1)

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 2.5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = Receipt()
    receipt.add_product(apples, 2.5, 1.99, 4.975)
    receipt.add_product(toothbrush, 1, 0.99, 0.99)

    receipt = teller.checks_out_articles_from(cart, receipt)

    verify(str(receipt))


def test_full_receipt_two_for_three():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.TWO_FOR_AMOUNT, apples, 3)

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 2.5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def test_full_receipt_two_for_one_one_item():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.TWO_FOR_AMOUNT, apples, 1)

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 1)
    cart.add_item_quantity(toothbrush, 1)

    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def test_full_receipt_five_for_one():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.FIVE_FOR_AMOUNT, apples, 1)

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 5)
    cart.add_item_quantity(toothbrush, 1)

    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def test_full_receipt_three_for_two_more_than_two_items():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.THREE_FOR_TWO, apples, "foo")

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 3)
    cart.add_item_quantity(toothbrush, 1)
    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def test_full_receipt_five_for_amount_less_than_five_items():
    apples, teller, toothbrush = create_teller_and_items()
    teller.add_special_offer(SpecialOfferType.FIVE_FOR_AMOUNT, apples, "foo")

    cart = ShoppingCart()
    cart.add_item_quantity(apples, 3)
    cart.add_item_quantity(toothbrush, 1)
    receipt = teller.checks_out_articles_from(cart)

    verify(str(receipt))


def create_teller_and_items():
    catalog = FakeCatalog()
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    catalog.add_product(toothbrush, 0.99)
    apples = Product("apples", ProductUnit.KILO)
    catalog.add_product(apples, 1.99)
    teller = Teller(catalog)
    return apples, teller, toothbrush
