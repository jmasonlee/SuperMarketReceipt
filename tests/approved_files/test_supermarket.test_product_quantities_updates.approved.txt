{apples  ProductUnit.KILO: 1.0, toothbrush  ProductUnit.EACH: 3.0}
            Receipt:
            
            Items:[apples  ProductUnit.KILO 1.0 1.99 1.99, toothbrush  ProductUnit.EACH 2.0 0.99 1.98, toothbrush  ProductUnit.EACH 1.0 0.99 0.99]
   
            
            Discounts:[]
   
            
            Total:4.96
   
        
        
